### PR TITLE
Added README.md for main with examples and explanations

### PR DESCRIPTION
**Pull request by DannyDaemonic**
I've been slowly working on this. It contains a couple of examples and a longer explanation of the options a user might use.

### CODE REVIEW

**Comment by DannyDaemonic** [comment]
Just updated it to fix a link to the primary README.

If you want to see it formatted, you can follow [this link](https://github.com/ggerganov/llama.cpp/blob/e7a09bdbabc354061829049ffa4d4a2cafcc84cc/examples/main/README.md).

**Comment by DannyDaemonic** [comment]
Thanks. That's the problem with this large write ups. We probably both started days ago...

**Comment by DannyDaemonic** [comment]
I was trying to compare all the options to see if we could merge the two. I did find his has instruction for `--seed` and `--ignore-eos`, which mine doesn't have and should certainly be in there. But our READMEs are structured so differently, I don't know if we could really merge them.

Mine is missing:
`-h, --help`
` -s SEED, --seed SEED`
`--ignore-eos`
`-b N, --batch_size N`

I did have `--batch_size` in there before but I accidentally cut it out of the final draft when I was restructuring things. I'm going to edit it back in again. I don't have any issue adding in the other missing options but I'll wait until we have a better idea of what we're doing.

**Comment by mgroeber9110** [comment]
@DannyDaemonic Sorry for the duplicate work - I should have probably made a note on the ticket that I intend to work on it, but your text definitely feels much more comprehensive, while I only took a quick stab at giving people a starting point.

From my point of view, it probably makes more sense to merge yours first, and then I can perhaps transfer a few tidbits of information from mine (such as what exactly ``--instruct`` injects, and perhaps a sentence or two on stopping conditions).

Just wondering: was it intentional to not mention ``--ignore-eos``? I found this quite powerful, but apparently it does not work everywhere (e.g. #990).

**Comment by DannyDaemonic** [comment]
Yeah, I feel bad that time was wasted either way.

The `--ignore-eos` was just an oversight. It really should be in there. I can make a stub for it in `## Generation Flags` and you can fill it in?

I intended to put all the options in there. Even `--help` is worth mentioning as it shows the latest options and default values, which can and do change frequently.

**Comment by DannyDaemonic** [comment]
I added the missing options with the exception of `--ignore-eos` since you have more experience with that than me. I was going to put a stub in there but it might not even be best under `## Generation Flags`. I was thinking maybe it belongs in `## Interaction` if that's how it's most commonly used? Either way, I could leave that up to you. Even if it does turn out to fit better in the generative section, you could certainly mention it in the instruct section.

On a bit of a tangent here, but I have had problems with Alpaca ending a response too soon, sometimes immediately, and never tried `--ignore-eos` to see if that helps. It does have a tendency to go a bit random at the end so I assumed it might make that worse. I suspect the real issue might be due to some early Alpaca models being trained with a context of 512 and I like to run everything with 2048.

File: examples/main/README.md [hunk 1]
```diff
@@ -1,3 +1,173 @@
...
+
+```bash
+./main -m models/7B/ggml-model.bin -n -1 --color -r "User:" --in-prefix " " --prompt $'User: Hi\nAI: Hello. I am an AI chatbot. Would you like to talk?\nUser: Sure!\nAI: What would you like to talk about?\nUser:'
+```
+
+## Common Options
+
+In this section, we cover the most commonly used options for running the `main` program with the LLaMA models:
+
+-   `-m FNAME, --model FNAME`: Specify the path to the LLaMA model file (e.g., `models/lamma-7B/ggml-model.bin`).
```
[the remark this answers is not in the document]
Comment (DannyDaemonic) [reply examples/main/README.md]: Sorry for marking this as resolved without replying, I thought I had replied.

I fixed the typo. Everything in common (except for model) is explained twice. The common section was just a way to show users the most frequently used options without overloading them. Sometimes people will see a block of text and give up (or go post a question instead), so those things are covered there briefly and then again a second time in more detail.

[the remark this answers is not in the document]
Comment (DannyDaemonic) [reply examples/main/README.md]: Also, once the threads issue gets resolved, I'll take threads out of the common section.

**Review by sw — APPROVED** [review]

File: examples/main/README.md [hunk 2]
```diff
@@ -1,3 +1,181 @@
...
+
+Example usage: `--top_p 0.9`
+
+By adjusting these options, you can control the diversity, quality, and creativity of the generated text to better suit your needs. You can experiment with different combinations of values to find the best settings for your specific use case.
+
+## Performance Tuning and Memory Options
+
+These options help improve the performance and memory usage of the LLaMA models:
+
+-   `-t N, --threads N`: Set the number of threads to use during computation. Using the correct number of threads can greatly improve performance. It is recommended to set this value to the number of CPU cores.
```
Comment (jon-chuang) [inline examples/main/README.md:163]: I'll say `physical CPU cores, and the number of performance cores in a chipset with Efficiency/Performance cores` piggybacked in PR https://github.com/ggerganov/llama.cpp/pull/934.

Comment (jon-chuang) [reply examples/main/README.md:163]: I still think we should leave it in; even with a warning the user might miss it; it's good to have multiple paths to getting the information.

**Pull request merged**